consensus/reactor.go

```diff
@@ -760,967 +662,688 @@ func (conR *Reactor) gossipVotesForHeight(
-func NewPeerState(peer p2p.Peer) *PeerState {
-	return &PeerState{
-		peer:   peer,
-		logger: log.NewNopLogger(),
-		PRS: cstypes.PeerRoundState{
-			Round:              -1,
-			ProposalPOLRound:   -1,
-			LastCommitRound:    -1,
-			CatchupCommitRound: -1,
-		},
-		Stats: &peerStateStats{},
-	}
-}
+	psKey := string(peerUpdate.PeerID)
 
-// SetLogger allows to set a logger on the peer state. Returns the peer state
-// itself.
-func (ps *PeerState) SetLogger(logger log.Logger) *PeerState {
-	ps.logger = logger
-	return ps
-}
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Do not allow starting new broadcasting goroutines after reactor shutdown
+		// has been initiated. This can happen after we've manually closed all
+		// peer goroutines and closed r.closeCh, but the router still sends in-flight
+		// peer updates.
+		if !r.IsRunning() {
+			return
+		}
 
-// GetRoundState returns an shallow copy of the PeerRoundState.
-// There's no point in mutating it since it won't change PeerState.
-func (ps *PeerState) GetRoundState() *cstypes.PeerRoundState {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
+		// save the peer if the ID is non-empty and we haven't seen this peer before
+		if psKey != "" && !r.peers.Has(psKey) {
+			peerState := NewPeerState(r.Logger, peerUpdate.PeerID)
+			r.peers.Set(psKey, peerState)
+		}
 
-	prs := ps.PRS // copy
-	return &prs
-}
+		peerState, ok := r.peers.Get(psKey).(*PeerState)
+		if !ok {
+			panic(fmt.Sprintf("peer %v has no state", psKey))
+		}
 
-// ToJSON returns a json of PeerState.
-func (ps *PeerState) ToJSON() ([]byte, error) {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
+		if !peerState.IsRunning() {
+			// Set the peer state's closer to signal to all spawned goroutines to exit
+			// when the peer is removed. We also set the running state to ensure we
+			// do not spawn multiple instances of the same goroutines and finally we
+			// set the waitgroup counter so we know when all goroutines have exited.
+			peerState.broadcastWG.Add(3)
+			peerState.SetRunning(true)
+
+			// start goroutines for this peer
+			go r.gossipDataRoutine(peerState)
+			go r.gossipVotesRoutine(peerState)
+			go r.queryMaj23Routine(peerState)
+
+			// Send our state to the peer. If we're fast-syncing, broadcast a
+			// RoundStepMessage later upon SwitchToConsensus().
+			if !r.WaitSync() {
+				r.sendNewRoundStepMessage(peerState.peerID)
+			}
+		}
 
-	return tmjson.Marshal(ps)
+	case p2p.PeerStatusDown, p2p.PeerStatusRemoved, p2p.PeerStatusBanned:
+		peerState, ok := r.peers.Get(psKey).(*PeerState)
+		if ok && peerState.IsRunning() {
+			// Close the closer to signal to all spawned goroutines to gracefully exit
+			// and we wait for them to all exit before marking the peer state as no
+			// longer running.
+			//
+			// XXX/TODO: It is possible that we can briefly wait here for all goroutines.
+			// Is this problematic?
+			peerState.closer.Close()
+			peerState.broadcastWG.Wait()
```

Comment:
This can wait for a long time, which will block `processPeerUpdate()` from processing further updates, potentially stalling the `PeerManager`. Might be better to just fetch the state and remove it from the map, then spawn a goroutine to stop the peerState.

Reply:
Yeah I've gone back-n-forth on this. This is the only bit of code I don't like. I'll marinate on your suggestion a bit 👍